patches/server/0936-Add-Entity-hidden-by-default-flag.patch

```diff
@@ -0,0 +1,207 @@
+-        }
+-        hidingPlugins.remove(CraftPlayer.getPluginWeakReference(plugin));
+-        if (!hidingPlugins.isEmpty()) {
+-            return; // Some other plugins still want the entity hidden
++        // Paper start - Entity hiddenByDefault
++        if (hidingPlugins != null) {
++            hidingPlugins.remove(CraftPlayer.getPluginWeakReference(plugin));
++            if (!hidingPlugins.isEmpty()) {
++                return; // Some other plugins still want the entity hidden
++            }
++        } else {
++            if (!entity.isHiddenByDefault()) {
++                return; // Entity isn't hidden
++            }
+         }
++        // Paper end
+         this.hiddenEntities.remove(entity.getUniqueId());
+ 
+         // Paper start
+@@ -1783,9 +1789,15 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+         }
+ 
+         ChunkMap.TrackedEntity entry = tracker.entityMap.get(other.getId());
+-        if (entry != null && !entry.seenBy.contains(this.getHandle().connection)) {
+-            entry.updatePlayer(this.getHandle());
++        // Paper start - Entity hiddenByDefault
++        if (entry != null) {
++            entry.showToEvenWhenHiddenByDefault.add(this.getHandle().connection);
++
++            if (!entry.seenBy.contains(this.getHandle().connection)) {
++                entry.updatePlayer(this.getHandle());
++            }
+         }
++        // Paper end
+     }
+     // Paper start
+     private void reregisterPlayer(ServerPlayer player) {
+@@ -1843,7 +1855,17 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+ 
+     @Override
+     public boolean canSee(org.bukkit.entity.Entity entity) {
+-        return !this.hiddenEntities.containsKey(entity.getUniqueId());
++        // Paper start - Entity hiddenByDefault
++        boolean ret = true;
++
++        if (entity.isHiddenByDefault()) {
++            ServerLevel world = ((CraftWorld) entity.getWorld()).getHandle();
++            ChunkMap.TrackedEntity entityTracker = world.getChunkSource().chunkMap.entityMap.get(entity.getEntityId());
++            ret = entityTracker.showToEvenWhenHiddenByDefault.contains(this.getHandle().connection);
++        }
++
++        // Paper end
++        return !this.hiddenEntities.containsKey(entity.getUniqueId()) && ret;
+     }
+ 
+     @Override
+@@ -2892,6 +2914,26 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+     }
+     // Paper end
+ 
++    // Paper start - Entity hiddenByDefault
++    @Override
++    public void setHiddenByDefault(boolean hiddenByDefault) {
++        // Update server lists
++        if (hiddenByDefault) {
++            server.getHandle().getPlayers().stream()
++                .filter(serverPlayer -> serverPlayer != this.getHandle())
++                .forEach(serverPlayer -> serverPlayer.connection.send(new ClientboundPlayerInfoPacket(ClientboundPlayerInfoPacket.Action.REMOVE_PLAYER, this.getHandle())));
```

Comment:
Any reason we don't check if the player can see the entity in the first place here ?

Reply:
The assumption is that, since we just called setHiddenByDefault, that any player will not be able to see the Entity since it has not been shown yet.

Reply:
But now that you mention it, would be good if the entity is already hidden. Will add a check.